.github/workflows/static-analysis.yaml

```
@@ -67,3 +67,52 @@ jobs:
       - name: Run pre-commit
         run: |
           pre-commit run --show-diff-on-failure --color=always --all-files
+
+  type-completeness-check:
+    name: Type completeness check
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+          fetch-depth: 0
+
+      - name: Set up uv
+        uses: astral-sh/setup-uv@v4
+        with:
+          python-version: "3.12"
+
+      - name: Calculate type completeness score
+        id: calculate_current_score
+        run: |
+          # `pyright` will exit with a non-zero status code if it finds any issues,
+          # so we need to explicitly ignore the exit code with `|| true`.
+          uv tool run --with . pyright --verifytypes prefect --ignoreexternal --outputjson > prefect-analysis.json || true
+          SCORE=$(jq -r '.typeCompleteness.completenessScore' prefect-analysis.json)
+          echo "current_score=$SCORE" >> $GITHUB_OUTPUT
+
+      - name: Checkout base branch
+        run: |
+          git checkout ${{ github.base_ref }}
+
+      - name: Calculate base branch score
+        id: calculate_base_score
+        run: |
+          uv tool run --with . pyright --verifytypes prefect --ignoreexternal --outputjson > prefect-analysis-base.json || true
+          BASE_SCORE=$(jq -r '.typeCompleteness.completenessScore' prefect-analysis-base.json)
+          echo "base_score=$BASE_SCORE" >> $GITHUB_OUTPUT
+
+      - name: Compare scores
+        run: |
+          CURRENT_SCORE=$(echo ${{ steps.calculate_current_score.outputs.current_score }})
+          BASE_SCORE=$(echo ${{ steps.calculate_base_score.outputs.base_score }})
+
+          if (( $(echo "$BASE_SCORE > $CURRENT_SCORE" | bc -l) )); then
+            echo "❌ Type completeness score has decreased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+            echo "Please add type annotations to your code to increase the type completeness score." >> $GITHUB_STEP_SUMMARY
+            exit 1
+          elif (( $(echo "$BASE_SCORE < $CURRENT_SCORE" | bc -l) )); then
+            echo "✅ Type completeness score has increased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+          else
+            echo "✅ Type completeness score remained unchanged at $BASE_SCORE" >> $GITHUB_STEP_SUMMARY
+          fi
```

would we want something like this?
```suggestion
          fi

    - name: Upload analysis artifacts
      uses: actions/upload-artifact@v4
      if: always()  # Upload even if the check fails
      with:
        name: type-completeness-reports
        path: |
          prefect-analysis.json
          prefect-analysis-base.json
```

---

Maybe? I could see that being useful if we wanted to look at historical data, and I'd probably only upload the current branch, but I can't think of a scenario where we'd want that. If we want the report, we can always generate it locally. 

Did you have a use case in mind that I'm missing?

---

this was the only thing I was thinking as well

> useful if we wanted to look at historical data

---

Gotcha, I think it's unlikely that we'll need that, but if we do, we can always checkout a commit and generate the `pyright` report on that commit.